### DISCUSSION TITLE
Using DNS

### DISCUSSION THREAD

**Discussion by jbbcst** [cesanta/mongoose#1693] [Ask questions, get answers] [answered]
Hello,

I am not using a Uri with an IP address but a name ( http://www.Name.com) to connect to a client. I guess I have to specify the ipv4 DNS, how and when should I specify it?

Thank you

**Comment by scaprile** [comment, marked as answer]
[mg_mgr_init()](https://mongoose.ws/documentation/#mg_mgr_init) will set Google DNS (8.8.8.8) as a default one.
If you want to set a specific DNS, just change that entry in the [mg_mgr](https://mongoose.ws/documentation/#struct-mg_mgr) structure
https://github.com/cesanta/mongoose/blob/0a265e79a67d7bfcdca27f2ccb98ccb474677ec6/mongoose.c#L3531

**Comment by jbbcst** [comment]
Hi, thank you for your response, I  am trying  to join a server as client, as you see below (capture of console debug), I changed the default DNS to tcp://192.168.0.126. I got an error but can't exactly know where it came from.
Did I miss something?Can we have more details about the error?

> 439891aa 2 main.cpp:86:main             Starting...
> 439891aa 3 mongoose.c:3479:mg_connect   1 -1 "https://ServerTest/api/Operation
> 439891aa 3 mongoose.c:3479:mg_connect   2 -1 tcp://192.168.0.126
> 439891aa 1 mongoose.c:404:mg_error      2 0xc8 connect: 10049
> 439895a2 1 mongoose.c:404:mg_error      1 0xffffffff DNS error
> 439895a2 3 mongoose.c:3456:mg_close_con 2 closed
> Close client connection
> 439895b1 3 mongoose.c:3456:mg_close_con 1 closed

**Reply by scaprile** [reply]
> tcp://192.168.0.126

`udp://8.8.8.8:53` means "connect to 8.8.8.8 using UDP at port 53". Most DNS servers use UDP. You specified TCP and did not specify the port number.

> 439895a2 1 mongoose.c:404:mg_error      1 0xffffffff DNS error

means "can't connect to the DNS server you told me to use using the configuration you told me to use"

Is there a DNS server running at 192.168.0.126 ? Does it use UDP or TCP ? At which port ? Is this reachable from the network where you have your Mongoose ? Then please properly configure the DNS URL with the correct address, protocol and port number; and if you still have problems please sniff your network using Wireshark or any other sniffer of your choice, and see the DNS traffic.

I am assuming your networking works and you already tried other examples as they are, and they work... otherwise I would ask you to do that first until you are sure your setup is OK before changing anything.

**Comment by jbbcst** [comment]
Thank you, you are right, I made a mistake. The DNS server use UDP and its port is 53.
I changed this to udp://192.168.0.126:53. I obtained this:

> 44c1c529 2 main.cpp:86:main             Starting...
> 44c1c529 3 mongoose.c:3479:mg_connect   1 -1 "https://ServerTest/api/Operation
> 44c1cd76 3 mongoose.c:3479:mg_connect   2 -1 udp://192.168.0.126:53
> 44c1cd76 3 mongoose.c:4123:mg_send      2 0xd0 0:0 24 err 0
> 44c1cda5 3 mongoose.c:4248:read_conn    2 0xd0 0:0 24 err 0
> 44c1cda5 1 mongoose.c:404:mg_error      1 0xffffffff  DNS lookup failed
> Close client connection
> 44c1cda5 3 mongoose.c:3456:mg_close_con 1 closed

I sniffed the network and compared with another application that trying to reach the same server, I am not a network master, however I saw this app trying to reach the server with its name ServerTest + the domain name as follows ServerTest.DomainName.
I changed the uri that I passed to mg_connect to _https://ServerTest.DomainName/api/Operation_, DNS seems to works but I can't reach the server:

> 44ce22a0 2 main.cpp:86:main             Starting...
> 44ce22a0 3 mongoose.c:3479:mg_connect   1 -1 "https://ServerTest.DomainName/api/Operation
> 44ce22b0 3 mongoose.c:3479:mg_connect   2 -1 udp://192.168.0.126:53
> 44ce22b0 3 mongoose.c:4123:mg_send      2 0xdc 0:0 38 err 0
> 44ce22cf 3 mongoose.c:4248:read_conn    2 0xdc 0:0 54 err 0
> 44ce22cf 3 mongoose.c:278:dns_cb        1 ServerTest.DomainName is 192.168.0.126
> 44ce22cf 1 mongoose.c:404:mg_error      1 0xa4 connect: 10049
> Close client connection
> 44ce22cf 3 mongoose.c:3456:mg_close_con 1 closed

For your information, the server I tried to reach has a self-signed certificate.

**Reply by scaprile** [reply]
Please read the tutorials and see the examples
https://mongoose.ws/tutorials/
https://mongoose.ws/tutorials/http-client/
First make it work without TLS, then enable TLS. Sniffing TLS connections is hard

**Comment by jbbcst** [comment]
Thanks, yes it works without TLS. I use a cmake project, the makefile is generated automatically. To activate TLS,  I saw in the tutorial that you have to compile with arguments.
Is this the only way to activate TLS?  ?

**Reply by scaprile** [reply]
Mongoose has to be built with MG_ENABLE_MBEDTLS=1 or MG_ENABLE_OPENSSL=1, depending on what TLS library you use.
Read your compiler/linker/IDE docs on how to enable symbols, or just hardcode a `#define` into `mongoose.h`

You have to tell your linker where to find your TLS library of choice. Read your compiler/linker/IDE docs.

I personally am null on cmake but Google surely can help

**Comment by cpq** [comment]
Maybe there is a tool CMakeMake already exists that generates CMakeLists.txt which in turn generates Makefile. The goal is to add `-DMG_ENABLE_MBEDTLS=1` to the compiler flags, as Sergio suggested above. I admit that is kinda complex, but I'm sure there should be tools that can do that.

**Comment by jbbcst** [comment]
Hello, thank you for your help,  I managed to enable TLS (as a reminder, the project works withoutTLS).I have the same result as in my last post with the mg_error connect: 10049. 
Is there anything to do when trying to connect to a server that has a self-signed certificate?

**Reply by scaprile** [reply]
If you sign your own certificates, you take the role of the CA, so you have to provide your CA certificate to those that need to validate your certificates... In other words: your `ca.pem` has to be installed at the client in order to validate the server certificate.
I don't remember error codes, add something like this to your event handler so we can see what that means:
```
  if (ev == MG_EV_ERROR) {
    MG_ERROR(("Error: %s", (char *) ev_data));
  }
```
Do you see in the logs that TLS is being started ? Are you doing what the tutorials say ? You should see a log line like this one when that happens:

> 1db78a26 3 tls_mbed.c:110:mg_tls_init   1 Setting TLS

**Comment by jbbcst** [comment]
Thanks, I added MG_ERROR(("Error: %s", (char *) ev_data)); but it gives the same error code without more details:

> 5eef0f40 1 main.cpp:74:fn               Error: connect: 10049

I don't see:

> 1db78a26 3 tls_mbed.c:110:mg_tls_init 1 Setting TLS

I added a breakpoint in the fn function and it doesn't enter in the block `else if (ev == MG_EV_CONNECT) {` but it passed successively in:
`if (ev == MG_EV_OPEN) {`
`else if (ev == MG_EV_ERROR) {`
`else if (ev == MG_EV_CLOSE)`

And yes my example is based on httpclient tutorial.

**Reply by scaprile** [reply]
OK, then it looks like you are not connecting... please sniff your network and see what is going on, this is at TCP level, maybe your server expects you to use a different port ?
ERROR and CLOSE are called when those conditions occur, and OPEN is called on start

**Comment by cpq** [comment]
@jbbcst to avoid wasting your and ours time, please publish your code. That's first. Second, you've asked a question and then you've got an answer. What follows, is not a discussion, but an issue, and does not belong here. Please file an issue and supply all required information, including the code you're using to connect.

**Comment by jbbcst** [comment]
Sorry to waste your time, I'm just trying to solve a technical point. Ok I will file an issue and supply all required information if I can't communicate with the server.

**Comment by jbbcst** [comment]
Hi, 
I have a  another question , the following logs show a connection as client to a server using TLS, there is a connection with id 8 and another with id 9. 
When I receive a response from the server I close manually the connection (with is_draining) so the connection with id 8 is closed.
However, the connection with id 9 is not closed, is this normal or should I do something to close it?

```
1f0a97f 3 mongoose.c:3479:mg_connect    8 -1 https://uriToConnect
1f0a980 3 mongoose.c:3479:mg_connect    9 -1 udp://192.xxxx.xxx:xx
1f0a980 3 mongoose.c:4123:mg_send       9 0x5 0:0 38 err 0
2022-09-21 17:30:11.944;DeviceService::onEScaleOperationValidatedXBS1
1f0acfe 3 mongoose.c:4248:read_conn     9 0x5 0:0 54 err 0
1f0acfe 3 mongoose.c:278:dns_cb         8 name.domain is 192.xxx.xxx.xxx
1f0acfe 3 mongoose.c:4325:mg_connect_re 8 0x6 -> c0a800c0:443 pend
1f0ad2c 3 mongoose.c:5268:mg_tls_init   8 Setting TLS, CA: null, cert: null, key: null
1f0ad2e 3 mongoose.c:5342:mg_tls_init   8 SSL client OK
1f0ad95 3 mongoose.c:5355:mg_tls_handsh 8 success
1f0ad95 3 mongoose.c:4258:write_conn    8 0x6 274:0 274 err 0
1f0adc9 3 mongoose.c:4248:read_conn     8 0x6 0:0 456 err 0
1f0adc9 3 mongoose.c:3456:mg_close_conn 8 closed
```

**Reply by scaprile** [reply]
If the last two 'x' on the second line correspond to what before your scrambling would have probably been a '5' and a '3', or whatever port number you have configured for your DNS, I would bet connection '9' is the DNS resolution attempt and has already been closed at 1f0acfe, probably.
Why do you say that connection is open ? Where do you see that ? You can increase verbosity level to see more fine grained operation if you wish/need to. Use MG_LL_VERBOSE (which happens to be 4)

**Comment by jbbcst** [comment]
Hi, I thought that because when I perform 2 communications (example below) it seems to keep the same communication opened (id 3 here) for the both. I don't see 3 closed even with verbosity level to 4. Maybe this behaviour is ok, just to know if I need to do something to explicitly close it.

```
c8cb0f 3 mongoose.c:3479:mg_connect     2 -1 https://servertest.domainname/api/Operation
c8cb0f 3 mongoose.c:3479:mg_connect     3 -1 udp://192.168.0.126:53
c8cb10 3 mongoose.c:4123:mg_send        3 0x5 0:0 38 err 0
c8cce4 3 mongoose.c:4248:read_conn      3 0x5 0:0 54 err 13
c8cce4 3 mongoose.c:278:dns_cb          2 ServerTest.DomainName is 192.168.0.125
c8cce5 3 mongoose.c:4325:mg_connect_res 2 0x3 -> c0a800c0:443 pend
c8cd17 3 mongoose.c:5268:mg_tls_init    2 Setting TLS, CA: null, cert: null, key: null
c8cd2a 3 mongoose.c:5342:mg_tls_init    2 SSL client OK
c8cd8e 3 mongoose.c:5355:mg_tls_handsha 2 success
c8cd8e 3 mongoose.c:4258:write_conn     2 0x3 274:0 274 err 0
c8cdc1 3 mongoose.c:4248:read_conn      2 0x3 0:0 456 err 0
c8cdc2 3 mongoose.c:3456:mg_close_conn  2 closed
```

Second communication:

```
c9701c 3 mongoose.c:3479:mg_connect     4 -1 https://servertest.domainname/api/Operation
c9701d 3 mongoose.c:4123:mg_send        3 0x5 0:0 38 err 0
c97047 3 mongoose.c:4248:read_conn      3 0x5 0:0 54 err 0
c97047 3 mongoose.c:278:dns_cb          4 ServerTest.DomainName is 192.168.0.125
c97047 3 mongoose.c:4325:mg_connect_res 4 0x4 -> c0a800c0:443 pend
c97072 3 mongoose.c:5268:mg_tls_init    4 Setting TLS, CA: null, cert: null, key: null
c97073 3 mongoose.c:5342:mg_tls_init    4 SSL client OK
c970d7 3 mongoose.c:5355:mg_tls_handsha 4 success
c970d8 3 mongoose.c:4258:write_conn     4 0x4 274:0 274 err 0
c9710b 3 mongoose.c:4248:read_conn      4 0x4 0:0 456 err 0
c9710b 3 mongoose.c:3456:mg_close_conn  4 closed
```

**Comment by scaprile** [comment]
No, that is the name resolution connection to your DNS. It is handled by Mongoose and you don't have to do anything about it.

**Reply by jbbcst** [reply]
ok thank you

**Comment by PythagoRascal** [comment]
I have a similar problem with a websocket client I'm trying to make. When I enable my VPN (PIA) I always get a DNS timeout. This seems to be because the forces DNS requests to go via it's own DNS server, which filters out direct DNS requests to a specific server. Interestingly enough, the same resolve works fine with `getaddrinfo` with the VPN enabled and I imagine it would also work when doing DNS-over-HTTPS.

I'd like to be able to distribute the client I'm trying to make in the future and because of this probably cannot know if they are behind a VPN or not. It would be really cool if there was an option to force Mongoose to use `getaddrinfo` for it's DNS resolves.

**Reply by cpq** [reply]
@PythagoRascal  `getaddrinfo()` is blocking, so it cannot be used.

You can update your app and query your OS about the DNS server IP address:
- on unixy systems, scan  `/etc/resolv.conf`
- on windowsy systems, read the registry `SYSTEM\\ControlSet001\\Services\\Tcpip\\Parameters\\Interfaces` key and fetch DNS IP from there, `NameServer` or `DhcpNameServer`

If that is successful, then set the DNS server URL in the manager:

```c
mg_mgr_init(&mgr);
if (dns_url != NULL) mgr.dns4.url = dns_url;  // dns_url should be "udp://IPADDR:53"
```

**Reply by PythagoRascal** [reply]
> `getaddrinfo()` is blocking, so it cannot be used.

Ah, darn, right.

I'll try out the options you mentioned. Thank you for the tip!

On another note: have you considered adding in DNS-over-HTTPS features to Mongoose? Or would that be too removed from the kind-of embedded-y nature of the library?

**Reply by cpq** [reply]
Yes, over HTTPS is just too much
